### PR TITLE
fix: Incorrect serialization of `IndexSignature` and `Record`

### LINKED ISSUE #5197: bug: Incorrect serialization of `IndexSignature` and `Record`
### Provide environment information

```
System:
  OS: Linux 6.2 Ubuntu 20.04.6 LTS (Focal Fossa)
  CPU: (4) x64 AMD EPYC 7763 64-Core Processor
  Memory: 11.48 GB / 15.61 GB
  Container: Yes
  Shell: 5.0.17 - /bin/bash
Binaries:
  Node: 20.10.0 - ~/nvm/current/bin/node
  Yarn: 1.22.19 - /usr/bin/yarn
  npm: 10.2.3 - ~/nvm/current/bin/npm
  pnpm: 8.12.1 - ~/.local/share/pnpm/pnpm
npmPackages:
  @trpc/client: link:packages/client => 10.44.1 
  @trpc/server: link:packages/server => 10.44.1 
  typescript: ^5.3.3 => 5.3.3 
```

### Describe the bug

Current implementation of `UndefinedToOptional` and `FilterDefinedKeys`  types causes incorrect type inference of the following: 

- `{ ... } & IndexSignature`
```ts
// Input
{
  a: number
  b: number
  [c: string]: number
}

// Output
{
  [x: number]: number | undefined;
  a: number;
  b: number;
}

// Expected
{
  [x: string]: number;
  [x: number]: number;
  a: number;
  b: number;
}
```

- `{ ... } & Record<K, T>`
```ts
// Input
{
  a: number
  b: number
} & Record<string, number>

// Output
{
  a: number
  b: number
}

// Expected
{
  [x: string]: number
  a: number
  b: number
}
```

- `... | undefined & IndexSignature | undefined`
```ts
// Input
{
  a: number
  b: number | undefined
  [c: string]: number | undefined
}

// Output
{
  [x: string]: number | undefined
  [x: number]: number | undefined
  a: number
}

// Expected (with "exactOptionalPropertyTypes": true)
{
  [x: string]: number
  [x: number]: number
  a: number
  b?: number
}
```

- `... | undefined & Record<K, T | undefined>`
```ts
// Input
{
  a: number
  b: number | undefined
} & Record<string, number | undefined>

// Output
{
  [x: string]: number | undefined
  a: number
}

// Expected (with "exactOptionalPropertyTypes": true)
{
  [x: string]: number
  a: number
  b?: number
}
```

With `"exactOptionalPropertyTypes": false (Default)`, the type `b?: number` changes to `b?: number | undefined`. 

### Link to reproduction

https://stackblitz.com/edit/github-2mjufe

### To reproduce

Apply the above cases in the following code and check the `Output` type.

```ts
import { inferRouterOutputs, initTRPC } from '@trpc/server';

type TOutput = Record<string, number> & {
  a: number;
  b: number;
};

const output = {
  a: 0,
  b: 1,
  c: 2,
} satisfies TOutput as TOutput;

const t = initTRPC.create();

const appRouter = t.router({
  output: t.procedure.query(() => output),
});

type AppRouter = typeof appRouter;
type AppRouterOutputs = inferRouterOutputs<AppRouter>;
type Output = AppRouterOutputs['output'];
```

### Additional information

I have come up with the following fix:

```diff
diff --git a/packages/server/src/shared/internal/serialize.ts b/packages/server/src/shared/internal/serialize.ts
index 64ce3890..402418ee 100644
--- a/packages/server/src/shared/internal/serialize.ts
+++ b/packages/server/src/shared/internal/serialize.ts
@@ -74,7 +74,7 @@ type FilterDefinedKeys<TObj extends object> = Exclude<
  * For an object T, if it has any properties that are a union with `undefined`,
  * make those into optional properties instead.
  *
- * Example: { a: string | undefined} --> { a?: string}
+ * Example: { a: string | undefined } --> { a?: string}
  */
 type UndefinedToOptional<T extends object> =
   // Property is not a union with `undefined`, keep as-is
@@ -82,6 +82,15 @@ type UndefinedToOptional<T extends object> =
     WithoutIndexSignature<T>,
     FilterDefinedKeys<WithoutIndexSignature<T>>
   > & {
-    // Property _is_ a union with `defined`. Set as optional (via `?`) and remove `undefined` from the union
-    [k in keyof Omit<T, FilterDefinedKeys<T>>]?: Exclude<T[k], undefined>;
+    // Property is a union with `defined`. Set as optional (via `?`) and remove `undefined` from the union
+    [k in keyof Omit<
+      WithoutIndexSignature<T>,
+      FilterDefinedKeys<WithoutIndexSignature<T>>
+    >]?: Exclude<T[k], undefined>;
+  } & {
+    // Finally add IndexSignature and remove `undefined` from the union
+    [k in keyof Omit<T, keyof WithoutIndexSignature<T>>]: Exclude<
+      T[k],
+      undefined
+    >;
   };
```

### 👨‍👧‍👦 Contributing

- [X] 🙋‍♂️ Yes, I'd be down to file a PR fixing this bug!

## Funding
* You can sponsor this specific effort via a [Polar.sh](https://polar.sh) pledge below
* We receive the pledge once the issue is completed & verified

<a href="https://polar.sh/trpc/trpc/issues/5197">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/trpc/trpc/issues/5197/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by microdev1**
This PR attempts to fix the following issue:

- Closes #5197

## 🎯 Changes

**Major**
- Changes the implementation of `UndefinedToOptional` type to support `IndexSignature` and `Record` types. All of the scenarios mentioned in the related issue (also defined in the `...test.ts` file) now produce the expected results.
- Generates opinionated types when `exactOptionalPropertyTypes=false`...
  -  ...and the input type contains an `undefined` or `optional` property and `IndexSignature` or `Record`. 
  - In this case, the `undefined` or `optional` property is merged into `IndexSignature` in the output type.

```ts
// exactOptionalPropertyTypes=false (Default)

// Input
{
  a: number
  b?: 0 | 1
  c: number | undefined
  [d: string]: number | undefined 
}

// Output (After `07a9922` fix)
// Produces incorrect type as `c?: number | undefined` isn't assignable to `[x: string]: number`
{
  [x: string]: number
  [x: number]: number
  a: number
  b?: 0 | 1 | undefined
  c?: number | undefined
}

// Output (After `a4a2653` opinionated change)
// The properties `b` and `c` are merged into `IndexSignature`
{
  [x: string]: number
  [x: number]: number
  a: number
}
```

```ts
// exactOptionalPropertyTypes=true

// Input
{
  a: number
  b?: 0 | 1
  c: number | undefined
  [d: string]: number | undefined 
}

// Output (Same for both `07a9922` and `a4a2653`)
// Attempts to preserve the input type
{
  [x: string]: number
  [x: number]: number
  a: number
  b?: 0 | 1
  c?: number
}
```

**Minor**
- Add implementation notes for types.
- Refactoring of `TObj` to `T` and `TKey`, `$Key`, `k` to `K` for consistency.

## ✅ Checklist

- [x] I have followed the steps listed in the [Contributing guide](https://github.com/trpc/trpc/blob/main/CONTRIBUTING.md).
- [x] If necessary, I have added documentation related to the changes made.
- [x] I have added or updated the tests related to the changes made.

### CODE REVIEW

File: packages/server/src/shared/internal/serialize.ts [hunk 1]
```diff
@@ -40,48 +40,83 @@ export type Serialize<T> =
...
- *
- * Example: { a: string | undefined} --> { a?: string}
+/**
+ * Get value of exactOptionalPropertyTypes config
+ */
+type ExactOptionalPropertyTypes = { a?: 0 | undefined } extends {
+  a?: 0;
+}
+  ? false
+  : true;
```
Comment (KATT) [inline packages/server/src/shared/internal/serialize.ts:84]: That's a funky one. Can we just assume "strictest" and make sure we document it in tRPC instead of having these hacks?

File: packages/tests/server/regression/issue-5197-output-with-index-signature-and-record.test.ts
Comment (KATT) [inline packages/tests/server/regression/issue-5197-output-with-index-signature-and-record.test.ts:1]: These tests could use the TS utility straight away rather than doing proper E2E-test - feels quite heavy-handed and hard to follow if this ever fails later :)

File: packages/server/src/shared/internal/serialize.ts (hunk 1 above)
Comment (microdev1) [reply packages/server/src/shared/internal/serialize.ts:84]: Hey, I agree about this being hacky, two options post removing it:
1. Allow incorrect types with `exactOptionalPropertyTypes=false`.
2. Generate merged type for both cases, looses type accuracy for literals when combined with `IndexSignature`.

In any case, I think a recommendation can be added in docs to use `exactOptionalPropertyTypes=true`. I tried turning it on in tRPC but the refactoring overhead is too much.

Let me know how to proceed and where to modify the docs. :)

Comment (KATT) [reply packages/server/src/shared/internal/serialize.ts:84]: > In any case, I think a recommendation can be added in docs to use exactOptionalPropertyTypes=true. I tried turning it on in tRPC but the refactoring overhead is too much.

Ohh, we should have strictest in tRPC too - could you open an issue and one of the core team will do it?

**Review by KATT — APPROVED** [review]
I'll merge this! Thanks!

**Comment by KATT** [comment]
The PR is pending to get merged, you can use it straight away afterwards by using the `@canary`-tag of tRPC which is deployed automatically from `main`:


Update your `package.json` to use the latest version of tRPC:

```diff
-    "@trpc/client": "^x.y.z",
-    "@trpc/next": "^x.y.z",
-    "@trpc/react-query": "^x.y.z",
-    "@trpc/server": "^x.y.z",
+    "@trpc/client": "canary",
+    "@trpc/next": "canary",
+    "@trpc/react-query": "canary",
+    "@trpc/server": "canary",
```

Then run `npm install` or similar. Reopen the issue if updating to the canary doesn't solve it.
